refactor git: Statement

alexLcmt opened the pull request.

alexLcmt committed:
refactor git: ContentBlock and Statement

jtranier commented on ui-components/src/components/player/ContentBlock.vue:
Pour régler le problème des composants qui doivent être nommés par 2 mots obligatoirement, et être homogène sur l'ensemble du code source, je te propose de nommer tous les composants spécifiques Elaastic en `ElXXX`.
Ici on va donc avoir `ElContentBlock`.

On va par contre réserver le préfixe `El` pour les spécifiques à Elaastic. On ne le fera pas pour `TipTap` par exemple.

alexLcmt replied:
J'ai modifié aussi le nom de Statement en ElStatement, c'est dans le commit [chore: rename ContentBlock and Statement],

jtranier commented on ui-components/src/components/player/ContentBlock.vue:
```diff
@@ -0,0 +1,99 @@
+  /**
+   * The title of the block.
+   */
+  title: {
```
Le titre doit être obligatoire.
Il faut ajouter `required: true`

alexLcmt replied:
J'ai modifié dans le commit [update: add required=true to title props in ContentBlock]

jtranier commented on ui-components/src/components/player/ContentBlock.vue:
```diff
@@ -0,0 +1,99 @@
+const openPanel = computed({
+  get: () => props.open ? [0] : [],
+  set: (val: number[] | number | null) => {
+    const isOpen = Array.isArray(val) ? val.includes(0) : val === 0;
```
Tu peux simplifier ça car le `v-expansion-panel` n'attend un tableau que lorsque l'on veut afficher plusieurs panneaux simultanément.
Dans notre cas il n'y en a qu'un, donc la valeurt `null` va le fermer et la valeur `0` va ouvrir le 1er et unique panneau

alexLcmt replied:
Oui c'est vrai, c'est plus simple. J'ai modifié dans le commit [update: modify openPanel in ContentBlock to make it simpler].

jtranier commented on ui-components/src/components/player/ContentBlock.vue:
```diff
@@ -0,0 +1,99 @@
+onMounted(() => {
+  if (!props.title || props.title.trim() === "") {
+    throw new Error("Prop 'title' is required and cannot be empty.");
+  }
```
Inutile à partir du moment où tu as mis le `required: true` dans la définition de la prop.
C'est VueJS qui gère ça.

alexLcmt replied:
J'aurais pu l'enlever directement quand j'ai rajouté le required: true mais j'avais pas vu, du coup j'ai fait un commit dédié : [update: remove onMounted function (replaced by required=true)]

jtranier commented on ui-components/src/components/statement/Statement.vue:
```diff
@@ -0,0 +1,153 @@
+   */
+  title: {
+    type: String,
+    default: "The title of this test statement"
```
Pas de valeur par défaut ici. Le `title` est requis.

alexLcmt replied:
J'ai enlevé la valeur par défaut et je l'ai remplacé par le required: true [update: remove default value for title in ElStatement and replace by required=true]

jtranier commented on ui-components/src/components/statement/Statement.vue:
```diff
@@ -0,0 +1,153 @@
+  /**
+   * The content of the statement, given as a string but actually is HTML code.
+   */
+  contentStatement: {
```
Je voudrais que cela passe par un `<slot>` et pas par une prop pour permettre de mettre n'importe quoi dedans.

alexLcmt replied:
Ok c'est noté, j'ai fait la modification dans [update: replace contentStatement props by a slot]

alexLcmt replied:
Mince j'ai oublié de mettre le v-if dans le slot, c'est corrigé dans ce commit : [update: change css style to vuetify classes and add v-if in the slot to make the hideStatement checkbox work]

jtranier commented on ui-components/src/components/statement/Statement.vue:
```diff
@@ -0,0 +1,153 @@
+  check3: {
+    type: Boolean,
+    default: false
+  },
```
En fait je ne comprends pas ce qu'est sensé faire ce composant.

alexLcmt replied:
En fait pour faire ce composant, je me suis appuyé sur la page Statement du player de test d'Elaastic (l'image jointe). Mais après réflexion c'est vrai que je ne vois nul part où ce composant apparait ou pourrait apparaitre. J'en déduis donc que je n'aurais pas dû refaire exactement ce qui se trouvait sur la page Statement du player de test d'Elaastic mais uniquement le panel. Les 3 checkboxs et le bouton n'étant là que pour illustrer les options que doit offrir ce composant. C'est à dire : ouvrir/fermer le panel, afficher/masquer le type de la question, afficher/masquer le contenu. 
![Capture_Statement_elaastic_player](https://github.com/user-attachments/assets/a3b0b2f6-f522-4044-ab82-968e521da6fa)

jtranier commented on ui-components/src/components/statement/Statement.vue:
```diff
@@ -0,0 +1,153 @@
+}
+.button:hover{
+  background-color: darkgray;
+}
```
Il faudrait utiliser les constructs de Vuetify plutôt que ce code CSS ad-hoc

alexLcmt replied:
J'ai remplacé le css par les classes vuetify et je me suis rendu compte que la checkbox hideStatement ne marchait pas car quand j'ai mis le slot à la place de la props, j'ai oublié de mettre le v-if, du coup j'ai aussi mis à jour. C'est dans ce commit : [update: change css style to vuetify classes and add v-if in the slot to make the hideStatement checkbox work]

jtranier commented on ui-components/src/stories/player/ContentBlock.stories.ts:
```diff
@@ -0,0 +1,104 @@
+  id: 1,
+  questionType: "MultipleChoice",
+  explanation: "",
+  choices: [],
```
Pourquoi créer une réponse sans choix ni explication ? C'est vraiment voulu ?
Et au-delà de ça, je suis entrain de me demander si on a vraiment besoin de créer une réponse pour illustrer le composant `ElContentBlock` ?

alexLcmt replied:
Ah oui ça c'est une erreur de ma part. En fait il y avait une story pour illustrer ElContentBlock avec le ResponseForm, mais étant donné que ce composant n'est pas présent dans cette branche j'ai supprimé la story mais j'ai complètement oublié de supprimer le code associé.

jtranier commented on ui-components/src/stories/player/ContentBlock.stories.ts:
```diff
@@ -0,0 +1,104 @@
+  questionType: "MultipleChoice",
+  explanation: "",
+  choices: [],
+  confidence: "Confiant(e)"
```
Le degré de confiance doit être une valeur numérique, ou une constante identifiant (voir ce qui est fait sur le serveur), mais pas un libellé en français.

jtranier replied:
@alexLcmt Tu n'a pas traité ce point

alexLcmt replied:
En fait ce code ne devrait pas être là, il sert à la story de ContentBlock qui intègre ResponseForm. Ce composant étant dans une autre branche, j'ai enlevé cette story (en tout cas pour cette branche). Mais je n'ai pas fait attention et ce code est resté. Je vais donc l'enlever et par contre je vais corriger cette story dans la branche feat/ui/response.

alexLcmt replied:
J'ai corrigé dans le commit [update: add ConfidenceDegree enum and use it instead of String for answers]

jtranier commented on ui-components/src/stories/player/ContentBlock.stories.ts:
```diff
@@ -0,0 +1,104 @@
+
+const validate = ref(false);
+
+function handleAnswer(newAnswer: AnyResponse) {
```
Je ne comprends pas pourquoi on a besoin de gérer des réponses pour illustrer ContentBlock.

jtranier replied:
Tu n'as pas répondu à cette question.
De plus si ce code est vraiment nécessaire, il faudra transformer les `answer` en `response`

alexLcmt replied:
La même chose qu'au-dessus. Ce code n'est pas nécessaire dans cette branche. Je vais l'enlever et faire la modification dans la branche feat/ui/response

alexLcmt replied:
J'ai corrigé dans le commit [update: replace answer by response in ContentBlock.stories]

jtranier commented on ui-components/src/stories/statement/Statement.stories.ts:
Pour l'instant je ne sais pas trop quoi dire sur cette story car j'ai besoin de mieux comprendre le rôle du composant qu'elle illustre

jtranier replied:
@alexLcmt Tu n'as pas répondu à ce point. 
Pour précier ma pensée, je pense que `ElStatement` qui instrumente `ElContentBlock` ça a du sens, mais je ne comprends pas pourquoi ce composant intègre des cases à cocher. Ces cases existent dans la page de test Elaastic. Elles ne doivent pas exister dans le composant. Avec storybook, c'est qui qui va créer des cases à cocher pour les prop de types Boolean.

alexLcmt replied:
Ok d'accord je comprend mieux. Effectivement je m'étais basé sur la page de test d'Elaastic. Je vais donc enlever les cases à cocher

alexLcmt replied:
J'ai corrigé dans le commit [update: remove checkbox from ElStatement]

alexLcmt committed:
chore: rename ContentBlock and Statement

alexLcmt committed:
update: add required=true to title props in ContentBlock

alexLcmt committed:
update: modify openPanel in ContentBlock to make it simpler

alexLcmt committed:
update: remove onMounted function (replaced by required=true)

alexLcmt committed:
update: remove default value for title in ElStatement and replace by required=true

alexLcmt committed:
update: replace contentStatement props by a slot

alexLcmt committed:
update: change css style to vuetify classes and add v-if in the slot to make the hideStatement checkbox work

alexLcmt commented:
Je n'ai pas encore renommer la branche statement en feat/ui/statement car j'avais peur que la PR disparaisse

jtranier commented:
@alexLcmt Tu as raison, à partir du moment où la PR est ouverte, il vaut mieux ne pas renommer la branche.

jtranier reviewed (changes requested):
Voir les commentaires qui ne sont toujours pas résolu

jtranier commented:
Ca serait bien de faire des stories pour illustrer toutes les situations de `ElContentBlock` (en fonction de ses props)

jtranier commented:
Quand on block n'est pas "collapsible", ce n'est pas compréhensible du point de vue l'utilisateur.
On ne devrait pas avoir le chevron qui sert à indiquer que l'on peut fermer.

jtranier commented:
Quand one ne cache pas le sous-titre & qu'il n'y a pas de sous-titre, on affiche juste des crochets

alexLcmt committed:
update: remove useless code from ContentBlock.stories

alexLcmt committed:
update: remove checkbox from ElStatement

alexLcmt committed:
update: panel icon doesn't appear when ContentBlock is not collapsible

alexLcmt committed:
update: add stories to ElContentBlock

alexLcmt commented:
J'ai ajouté des stories à ElContentBlock dans [update: add stories to ElContentBlock]. Pour les crochets des subtitle c'était déjà le cas mais il n'y avait pas de story qui le montrait, j'en ai ajouté une dans le même commit. J'ai enlevé l'icon de ElContentBlock quand il n'est pas collapsible dans le commit [update: panel icon doesn't appear when ContentBlock is not collapsible].

jtranier committed:
chore:  Refactor subtitle prop and improve code consistency

Renamed `isSubtitleHidden` to `showSubtitle` for better readability.
Do not show the subtitle when it's empty (which only display empty brackets)

jtranier committed:
chore: Set ElContentBlock collapsible by default

jtranier committed:
chore: Code improving using v-model + rework default story

jtranier committed:
chore: Simplify ElContentBlock.stories.ts

jtranier committed:
chore: Add more meaningful content to ElContentBlock story

jtranier committed:
fix: Subtitle position

jtranier committed:
feat: Remove "showSubtitle" prop

jtranier committed:
chore: Reindent ElStatement.vue

jtranier committed:
chore: Remove ElStatement

It brings nothing more than ElContent.

jtranier committed:
feat: Add ElContentBlock Statement story

jtranier committed:
feat: Improve style of ElContent title

jtranier committed:
feat: Add ElHtmlContent component to present HTML content

jtranier committed:
fix: Do not rename component within templates

jtranier committed:
fix: Remove "update:isSubtitleHidden" event

jtranier committed:
feat: Add "showContent" prop on ElContentBlock.vue
